### ISSUE TITLE
customComponent with ngModal in ngFor with ngIf not work

### ISSUE THREAD

**Issue by Willyisback** [angular/angular#33626]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/forms or @angular/core. 

### Is this a regression?

Yes, the previous version in which this bug was not present was: 8

### Description

I have updated my project to Angular 9. Now my custom component, where ControlValueAccessor is implemented, doesn't work in ngFor anymore. I get the following error: ` Property 'demoItem' does not exist on type 'AppComponent'.`

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-issue-repro2-h9tqtj?file=src%2Fapp%2Fdemo-item.component.ts

The Stackblitz example unfortunately works. But if I want to start the same project locally, it doesn't work. So I push my projekt to GitHub: https://github.com/Willyisback/angular-bug-testing

## 🔥 Exception or Error
<pre><code>

ERROR in src/app/app.component.html:2:71 - error TS2339: Property 'demoItem' does not exist on type 'AppComponent'.

2   <app-demo-item *ngIf="demoItem?.type === '1'" [(ngModel)]="demoItem"></app-demo-item>
                                                                        ~~~~~~

  src/app/app.component.ts:11:16
    11   templateUrl: './app.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.
src/app/app.component.html:3:71 - error TS2339: Property 'demoItem' does not exist on type 'AppComponent'.

3   <app-demo-item *ngIf="demoItem?.type === '2'" [(ngModel)]="demoItem"></app-demo-item>
                                                                        ~~~~~~

  src/app/app.component.ts:11:16
    11   templateUrl: './app.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.
src/app/app.component.html:4:71 - error TS2339: Property 'demoItem' does not exist on type 'AppComponent'.

4   <app-demo-item *ngIf="demoItem?.type === '3'" [(ngModel)]="demoItem"></app-demo-item>
                                                                        ~~~~~~

  src/app/app.component.ts:11:16
    11   templateUrl: './app.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 9.0.0-rc.0
Node: 12.13.0
OS: linux x64
Angular: 9.0.0-rc.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.0
@angular-devkit/build-angular     0.900.0-rc.0
@angular-devkit/build-optimizer   0.900.0-rc.0
@angular-devkit/build-webpack     0.900.0-rc.0
@angular-devkit/core              9.0.0-rc.0
@angular-devkit/schematics        9.0.0-rc.0
@ngtools/webpack                  9.0.0-rc.0
@schematics/angular               9.0.0-rc.0
@schematics/update                0.900.0-rc.0
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2

</code></pre>

**Anything else relevant?**

**Comment by alexzuza** [comment]
According to [Ivy Compatibility Guide](https://next.angular.io/guide/ivy-compatibility)

> It is now an error to assign values to template-only variables like item in ngFor="let item of items" (previously, the compiler would ignore these assignments).

Minimal reproduction is:

```
<div *ngFor="let item of [1, 2]">
  <div (click)="item = 3;"></div>
</div>
```
In your example you're doing almost the same thing:

```
<div *ngFor="let demoItem of demo">
  <...  [(ngModel)]="demoItem"> => (ngModelChange)="demoItem = $event"
```

**Comment by kara** [comment]
This is an expected breaking change, but looks like template type-checking isn't doing the right thing here. We should be throwing a friendly error.

**Comment by alxhub** [comment]
This was fixed by #34339 🎉
